Published Dashboard updates, toc performance plot improvements and ts cleaning functions

SamStruthers opened the pull request:
Hi @juandlt-csu 

Sorry for another big (and delayed PR). Most of it I think you can ignore as it is just code to create/modify plots. 
Scripts with mostly plotting changes:
- `modeling/toc/03_toc_model.rmd`
- `modeling/toc/04_apply_toc_model.rmd`
- `scratch/ROSS_FC_comparisons.rmd`
- `dashboard/server.R` Lines greater than 387
Server code: If you want to see more shiny code, feel free to check it out. Otherwise, it works well enough on the published version and closes #22 & closes #28  (for our alpha version)

Mostly I am curious to get your thoughts on these four functions:

- `src/apply_cleaning_filters.R`: cleans up the value column based on flags and deals with sites/parameters/flags where we are overflagging

- `src/apply_linear_interpolation_missing_data.R`: If there are NAs in the value column, we perform a linear interpolation between existing data and only fill gaps of a given size (1 hour in our case)

- `src/apply_low_pass_binominal_filter.R`: Very similar to your `fcw.qaqc` low pass filter but adds in original data if smoothed mean could not be computed (ie missing data/edges)

- `src/apply_timestep_median.R`: computes median of data for a given timestep

These functions are used one after the other in the `server.R` file from lines 313-326 to clean up our live data. My goal was to make these easy to apply to other datasets (ie why users can add their own site_col, parameter_col, etc) and perform our cleaning steps in a tidy manner (not 100 lines of code)

juandlt-csu commented on dashboard/R/apply_low_pass_binomial_filter.R:
```diff
@@ -0,0 +1,174 @@
+#' current measurement, moderate influence to the measurements 15 minutes before and
+#' after, and least influence to the measurements 30 minutes before and after. The
+#' triple application of the filter creates a more aggressive smoothing effect equivalent
+#' to a higher-order binomial filter.
```
thought: though it is mentioned here it is worth noting that the equivalent binomial filter becomes `[1,12,66,220,495,792,924,792,495,220,66,12,1]/4096`. This means that we are taking more than just the 15/30 minutes before and after into consideration for the final smoothed value when we do this. The influence of those further away values is very small, but still there.

SamStruthers replied:
I'll open an issue for us to discuss at next working meeting!

juandlt-csu commented on dashboard/R/apply_low_pass_binomial_filter.R:
```diff
@@ -0,0 +1,174 @@
+#' Default is "dataframe".
+#' @param site_col The name of the column containing site identifiers. Default is "site".
+#' @param parameter_col The name of the column containing parameter identifiers. Default is "parameter".
+#' @param time_col The name of the column containing datetime information. Must be a `POSIXct` or numeric type, or otherwise convertible to numeric. Default is "DT_round".
```
nitpick: we should change `time_col` to `dt_col`.

juandlt-csu commented on dashboard/R/apply_low_pass_binomial_filter.R:
```diff
@@ -0,0 +1,174 @@
+            mutate(!!sym(new_value_col) := NA_real_)
+        }
+        # Fill remaining NA in smoothed column with original values
+        .x <- .x %>%
+          mutate(
+            !!sym(new_value_col) := if_else(
+              is.na(!!sym(new_value_col)) & !is.na(!!sym(value_col)),
+              !!sym(value_col),
+              !!sym(new_value_col)
+            )
+          )
```
How was trying to come up with a algorithmic approach to this instead of filling in with og values? I'm curious to understand the reasoning behind interacting with NA values for this filter. Even when you do "actual" fourier transform low pass filters it doesn't like NA values.

SamStruthers replied:
Added to issue

juandlt-csu commented on dashboard/R/apply_linear_interpolation_missing_data.R:
```diff
@@ -0,0 +1,70 @@
+#'   identifiers.
+#' @param parameter_col A string giving the name of the column containing
+#'   parameter identifiers.
+#' @param time_col A string giving the name of the column containing datetime
+#'   information. Must be a `POSIXct` or numeric type, or otherwise convertible
+#'   to numeric.
```
suggestion: change `time_col` to `dt_col`

juandlt-csu commented on dashboard/R/apply_linear_interpolation_missing_data.R:
```diff
@@ -0,0 +1,70 @@
+#'   gap-filled values. Default is `"mean_filled"`. If this matches `value_col`,
+#'   the original values will be overwritten.
+#'
+#' @return A data frame with the original columns plus an additional column
+#'   containing the linearly interpolated values (or with the original column
+#'   replaced if `new_value_col` = `value_col`).
```
suggestion: This seems dubious to me, I think that it should not overwrite the original column.

juandlt-csu replied:
I could go either way on this though.

SamStruthers replied:
Added a note that this is possible but not recommended

juandlt-csu commented on dashboard/R/apply_linear_interpolation_missing_data.R:
```diff
@@ -0,0 +1,70 @@
+#' Gap fill short periods of missing df with linear interpolation
```
Will this be used for everything not just turb?

SamStruthers replied:
yup, this is for all parameters. This is primarily to fill areas where we removed outliers in QAQC or if data was missing for a short period of time (onboard/telemetry malfunction)

juandlt-csu commented on dashboard/R/apply_linear_interpolation_missing_data.R:
```diff
@@ -0,0 +1,70 @@
+    arrange(!!sym(site_col), !!sym(parameter_col), !!sym(time_col)) %>%
+    group_by(!!sym(site_col), !!sym(parameter_col)) %>%
+    mutate(
+      !!sym(new_value_col) := zoo::na.approx(
```
question: did you play around with `na.spline` at all? Linear interpolation is fine for this but I'm curious if there were other options you explored.

juandlt-csu commented on dashboard/R/apply_cleaning_filters.R:
This is a great function for letting us know what our QAQC flags are doing wrong, haha.

juandlt-csu commented on dashboard/R/apply_timestep_median.R:
```diff
@@ -0,0 +1,104 @@
+#'
+#' @param df Input data containing water quality measurements. Must include columns for
+#' site, datetime, parameter, and value.
+#' @param data_format The format of the input data. Options are either `"dataframe"` or `"list"`.
+#' If `"dataframe"` is selected, all data (multiple site-parameter combinations) are stored in
+#' a single long dataframe. If `"list"` is selected, data are stored in a list of dataframes,
+#' with each dataframe containing data for a single site-parameter combination.
+#' Default is `"dataframe"`.
```
Same thing as the smoothing function, this should just take in dfs.

juandlt-csu commented on dashboard/R/apply_timestep_median.R:
Generally the same notes that I gave on the low pass filter option regarding design.

juandlt-csu reviewed (approved):
Hey Sam, I think that these functions do what you want, but I commented on some design choices that I think could make these functions more usable to work with and build upon in the future. Since these are design changes I can make those as the "dev" guy on this team. `apply_cleaning_filters` is a great way to see where our qaqc falls short. I am curious about playing around with options/algorithms to deal with NAs in the binomial filter. Overall I think this works.

juandlt-csu commented on dashboard/server.R:
```diff
@@ -267,75 +282,65 @@ server <- function(input, output, session) {
+    req(base_filtered_data())
+
+    if (isTRUE(input$apply_qaqc_filter)) {
+      #remove flagged data
+      cleaned_data <- apply_cleaning_filters(df = base_filtered_data(), new_value_col = "mean_cleaned")
+      # interpolate missing data (<1 hour)
+      filled_data  <- apply_linear_interpolation_missing_data(df = cleaned_data, value_col = "mean_cleaned", time_col = "DT_round_MT")
+      # apply binomial filter to turbidity and Chl-a
+      smoothed_data <- apply_low_pass_binomial_filter(df = filled_data, value_col = "mean_filled", new_value_col = "mean_smoothed", time_col = "DT_round_MT")
+      # apply timestep median to reduce noise
+      apply_timestep_median(df = smoothed_data, value_col = "mean_smoothed", new_value_col = "mean", timestep = input$data_timestep, time_col = "DT_round_MT")
```
To do: You should pipe these so that all these objects are not stored in memory in the app. Efficiency gain of not having to store/keep track of 4 very similar datasets. I see that this is some server.R stuff that maybe I don't understand, but unless you need to call/use these dfs outside of this reactive() call I suggest you pipe these. If that is just not how this stuff works then my bad, you can ignore this.

SamStruthers commented on dashboard/R/apply_low_pass_binomial_filter.R:
```diff
@@ -0,0 +1,174 @@
+
+    return(data_smoothed)
+  } else {
+    stop("Invalid data_format. Choose either 'dataframe' or 'list'.")
```
will do!

SamStruthers commented on dashboard/R/apply_linear_interpolation_missing_data.R:
```diff
@@ -0,0 +1,70 @@
+#' long missing periods.
+#'
+#' @param df A data frame containing time series data with missing values
+#'   represented as `NA`. Data should be padded to regular time intervals (e.g.,
+#'   using [padr::pad()]) before applying this function.
```
Added an attempt at a check, but if you may still want to check it out

SamStruthers commented:
Hi Juan, 
Fixed the majority of your comments above and added an issue for our low pass filter discussion. I cleaned up that server code you mentioned. I also added my `calc_lagged_parameters` function if you want to take a peek at that

juandlt-csu commented on src/calc_lagged_parameter.R:
```diff
@@ -0,0 +1,88 @@
+      !!sym(new_value_col) := slide_period_dbl(
+        .x = !!sym(parameter_col), # use the specified parameter column
+        .i = !!sym(dt_col), # use the time column for indexing
+        .period = time_period_split[2], # use the specified time unit
```
nitpick: `time_period_num` is created from `time_period_split`, you could make a `time_period_str` that has this string from `time_period_split` for consistency here.

juandlt-csu reviewed (approved):
Hey, Sam! I reviewed your lag calculation function. I saw some documentation inconsistencies and I had some nitpicks, but overall it looks like it should do what you expect it to. Very cool implementation of `switch()`.

SamStruthers commented:
Thanks for checking that over Juan! I had copied over the documentation from the other function and forgot to push my commit that had corrections for the docs. I also added a quick change to the parsing of `time_period` which I think is a bit cleaner now.
